### PR TITLE
test: Add integration tests for downscoping

### PR DESCRIPTION
**Pull request by bojeil-google**
This tests creates a temporary bucket and 2 objects in it.
A downscoped token is then created to access only one of the objects (readonly).
The test would then check:
- Read access to accessible object is successful.
- Write access to that object is unsuccessful.
- Read access to the inaccessible object is not successful.

### CODE REVIEW

**Comment by tseaver** [comment]
@bojeil-google Unfortunately, your PR has landed in a pocket where the system tests are not being run on Kokoro.  @dandhlee is working to correct that issue (see #836).  Please hold off merging here until `master` has those tests running / passing in the `Kokoro system-3.7` job, and then merge `master` to your branch.

File: system_tests/noxfile.py [hunk 1]
```diff
@@ -350,6 +350,20 @@ def external_accounts(session):
...
+        "google-auth",
+        "google-cloud-storage",
+    )
+    default(
+        session,
+        "system_tests_sync/test_downscoping.py",
+        *session.posargs,
+    )
+
+
```
Comment (tseaver) [inline system_tests/noxfile.py:366]: Do we need to add a parallel `async` version of this test?

**Comment by dandhlee** [comment]
@tseaver The system test does run, but it just doesn't show up on the dashboard here. I can see on the log that both the `presubmit` and `system-3.7` has run successfully.

File: system_tests/system_tests_sync/test_downscoping.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+# The object name of the object inaccessible by the downscoped token.
+_INACCESSIBLE_OBJECT_NAME = "other-customer-data.txt"
+
+
+@pytest.fixture
+def temp_bucket():
+    """Yields a bucket that is deleted after the test completes."""
+    bucket = None
+    while bucket is None or bucket.exists():
+        bucket_name = "bucket-downscoping-test-{}".format(uuid.uuid4())
```
Comment (tseaver) [inline system_tests/system_tests_sync/test_downscoping.py:45]: Can we use a prefix which indicates that this repo's tests own the bucket?  That will make chasing down any possible leak sources simpler. E.g.:
```suggestion
        bucket_name = "google-auth-python-downscoping-test-{}".format(uuid.uuid4())
```

File: system_tests/system_tests_sync/test_downscoping.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        bucket = storage.Client().bucket(bucket_name)
+    bucket = storage.Client().create_bucket(bucket.name)
+    yield bucket
+    bucket.delete(force=True)
+
+
+@pytest.fixture
+def temp_blobs(temp_bucket):
+    """Yields a blob that is deleted after the test completes."""
```
Comment (tseaver) [inline system_tests/system_tests_sync/test_downscoping.py:54]:
```suggestion
    """Yields two blobs that are deleted after the test completes."""
```

File: system_tests/system_tests_sync/test_downscoping.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    bucket = temp_bucket
+    # Downscoped tokens will have readonly access to this blob.
+    accessible_blob = bucket.blob(_ACCESSIBLE_OBJECT_NAME)
+    accessible_blob.upload_from_string(_ACCESSIBLE_CONTENT)
+    # Downscoped tokens will have no access to this blob.
+    inaccessible_blob = bucket.blob(_INACCESSIBLE_OBJECT_NAME)
+    inaccessible_blob.upload_from_string(_INACCESSIBLE_CONTENT)
+    yield (accessible_blob, inaccessible_blob)
```
Comment (tseaver) [inline system_tests/system_tests_sync/test_downscoping.py:62]: We promise above to delete on exit:
```suggestion
    yield (accessible_blob, inaccessible_blob)
    bucket.delete_blobs([accessible_blob, inaccessible_blob])
```

File: system_tests/system_tests_sync/test_downscoping.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    """
+    # Initialize the Credential Access Boundary rules.
+    available_resource = f"//storage.googleapis.com/projects/_/buckets/{bucket_name}"
+    # Downscoped credentials will have readonly access to the resource.
+    available_permissions = ["inRole:roles/storage.objectViewer"]
+    # Only objects starting with the specified prefix string in the object name
+    # will be allowed read access.
+    availability_expression = (
+        "resource.name.startsWith('projects/_/buckets/{}/objects/{}')".format(
+            bucket_name, object_prefix
```
Comment (tseaver) [inline system_tests/system_tests_sync/test_downscoping.py:85]:
```suggestion
        f"resource.name.startsWith('projects/_/buckets/{bucket_name}/objects/{object_prefix}')"
```

File: system_tests/system_tests_sync/test_downscoping.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+_ACCESSIBLE_OBJECT_NAME = f"{_OBJECT_PREFIX}-data.txt"
+# The content of the object accessible by the downscoped token.
+_ACCESSIBLE_CONTENT = "hello world"
+# The content of the object inaccessible by the downscoped token.
+_INACCESSIBLE_CONTENT = "secret content"
+# The object name of the object inaccessible by the downscoped token.
+_INACCESSIBLE_OBJECT_NAME = "other-customer-data.txt"
+
+
+@pytest.fixture
```
Comment (tseaver) [inline system_tests/system_tests_sync/test_downscoping.py:40]: Prefer an explicit scope for fixtures, e.g.:
```suggestion
@pytest.fixture(scope="module")
```

File: system_tests/system_tests_sync/test_downscoping.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    bucket = None
+    while bucket is None or bucket.exists():
+        bucket_name = "bucket-downscoping-test-{}".format(uuid.uuid4())
+        bucket = storage.Client().bucket(bucket_name)
+    bucket = storage.Client().create_bucket(bucket.name)
+    yield bucket
+    bucket.delete(force=True)
+
+
+@pytest.fixture
```
Comment (tseaver) [inline system_tests/system_tests_sync/test_downscoping.py:52]: Likewise here:
```suggestion
@pytest.fixture(scope="module")
```

File: system_tests/system_tests_sync/test_downscoping.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    # Initialize a Cloud Storage client with the oauth2 credentials.
+    storage_client = storage.Client(credentials=creds)
+
+    # Test read access succeeds to accessible blob.
+    bucket = storage_client.bucket(bucket_name)
+    blob = bucket.blob(accessible_blob.name)
+    assert blob.download_as_bytes().decode("utf-8") == _ACCESSIBLE_CONTENT
+
+    # Test write access fails.
+    with pytest.raises(exceptions.GoogleCloudError) as excinfo:
```
Comment (tseaver) [inline system_tests/system_tests_sync/test_downscoping.py:154]: This error should likely be more strongly typed, e.g.:
```suggestion
    with pytest.raises(exceptions.Unauthorized) as excinfo:
```
or:
```suggestion
    with pytest.raises(exceptions.Forbidden) as excinfo:
```

**Review by arithmetic1728 — APPROVED** [review]

File: system_tests/system_tests_sync/test_downscoping.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        blob.upload_from_string("Write operations are not allowed")
+
+    assert excinfo.match(r"does not have storage.objects.create access")
+
+    # Test read access fails to inaccessible blob.
+    with pytest.raises(exceptions.GoogleCloudError) as excinfo:
```
Comment (tseaver) [inline system_tests/system_tests_sync/test_downscoping.py:160]: LIkewise here:
```suggestion
    with pytest.raises(exceptions.Unauthorized) as excinfo:
```
or:
```suggestion
    with pytest.raises(exceptions.Forbidden) as excinfo:
```

**Review by tseaver — CHANGES_REQUESTED** [review]

**Comment by arithmetic1728** [comment]
@bojeil-google please take a look at @tseaver 's comments

File: system_tests/noxfile.py (hunk 1 above)
Comment (bojeil-google) [reply system_tests/noxfile.py:366]: I am not sure about the `async` project and effort. My team is not involved in that. I would rather not make this a blocking requirement for this project. I will get more information on that and how to resolve that long term.

File: system_tests/system_tests_sync/test_downscoping.py (hunk 2 above)
Comment (bojeil-google) [reply system_tests/system_tests_sync/test_downscoping.py:45]: I updated the name but there is a size limit and I seem to be getting some errors when using `google` in the name.
`Use of this bucket name is restricted: 'goo..`
I removed that from the name.

File: system_tests/system_tests_sync/test_downscoping.py (hunk 3 above)
Comment (bojeil-google) [reply system_tests/system_tests_sync/test_downscoping.py:54]: Done

File: system_tests/system_tests_sync/test_downscoping.py (hunk 4 above)
Comment (bojeil-google) [reply system_tests/system_tests_sync/test_downscoping.py:62]: Done.

File: system_tests/system_tests_sync/test_downscoping.py (hunk 5 above)
Comment (bojeil-google) [reply system_tests/system_tests_sync/test_downscoping.py:85]: Done

File: system_tests/system_tests_sync/test_downscoping.py (hunk 6 above)
Comment (bojeil-google) [reply system_tests/system_tests_sync/test_downscoping.py:40]: Done

File: system_tests/system_tests_sync/test_downscoping.py (hunk 7 above)
Comment (bojeil-google) [reply system_tests/system_tests_sync/test_downscoping.py:52]: Done

File: system_tests/system_tests_sync/test_downscoping.py (hunk 8 above)
Comment (bojeil-google) [reply system_tests/system_tests_sync/test_downscoping.py:154]: Done

File: system_tests/system_tests_sync/test_downscoping.py (hunk 9 above)
Comment (bojeil-google) [reply system_tests/system_tests_sync/test_downscoping.py:160]: Done

**Comment by bojeil-google** [comment]
All comments were addressed in https://github.com/googleapis/google-auth-library-python/pull/838

**Pull request merged**